BUG: preserve EA dtype in transpose

TomAugspurger opened the pull request.

TomAugspurger committed:
BUG: preserve EA dtype in transpose

jbrockmendel commented:
does this subsume #28048?

jbrockmendel commented on doc/source/whatsnew/v1.0.0.rst:
```diff
@@ -775,6 +775,7 @@ Reshaping
 - Bug where :meth:`DataFrame.equals` returned True incorrectly in some cases when two DataFrames had the same columns in different orders (:issue:`28839`)
 - Bug in :meth:`DataFrame.replace` that caused non-numeric replacer's dtype not respected (:issue:`26632`)
 - Bug in :func:`melt` where supplying mixed strings and numeric values for ``id_vars`` or ``value_vars`` would incorrectly raise a ``ValueError`` (:issue:`29718`)
+- Dtypes are now preserved when transposing a ``DataFrame`` where each column is the same extension dtyep (:issue:``)
```
dtyep -> dtype

TomAugspurger committed:
fix typo

TomAugspurger committed:
remove xpass

TomAugspurger commented:
> does this subsume #28048?

Ah, yeah forgot about that one.

jbrockmendel commented:
Easy to forget about, i closed it to clear the queue and then only reopened it a couple days ago.  I'm going to re-close it in favor or this one; maybe its worth salvaging some of the tests it implemented

TomAugspurger commented on pandas/core/generic.py:
```diff
@@ -721,12 +721,37 @@ def transpose(self, *args, **kwargs):
+
+            # Slow, but unavoidable with 1D EAs.
+            new_values = []
+            for i in range(len(self)):
```
I'm rethinking this approach. This results in `n_rows * n_columns` `__getitem__`s. My intent was to avoid going through a 2D object-dtype ndarray. But we're essentially doing that with lists. So I think it'll be better to just do `.values.T` and then rebuild the EAs from the object-dtype array.

jreback commented on pandas/core/generic.py:
```diff
@@ -721,12 +721,37 @@ def transpose(self, *args, **kwargs):
+            kwargs.pop("copy", None)  # by definition, we're copying
+            dtype = self._data.blocks[0].dtype
+            arr_type = dtype.construct_array_type()
+
```
I would move this logic to pandas/core/reshape/reshape.py this has a lot of similiarity to _unstack_extension_series

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -721,12 +721,37 @@ def transpose(self, *args, **kwargs):
+        if (
+            self._is_homogeneous_type
+            and len(self._data.blocks)
+            and is_extension_array_dtype(self._data.blocks[0].dtype)
```
we can avoid self._data references by making this `len(self.dtypes) and is_extension_array_dtype(self.dtypes.iloc[0])`

jbrockmendel replied:
ditto on 731 with self.dtypes

TomAugspurger committed:
simplify

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-transpose

TomAugspurger committed:
simplify

TomAugspurger committed:
steal tests

TomAugspurger committed:
update docs

TomAugspurger commented:
Pushed a largeish refactor. We actually don't need NDFrame.transpose anymore. Series gets its from IndexOpsMixin. So I moved the logic to DataFrame.transpose, and was able to remove all the axes handling stuff.

> maybe its worth salvaging some of the tests it implemented

Added.

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -2596,7 +2601,27 @@ def transpose(self, *args, **kwargs):
+        if self._is_homogeneous_type and is_extension_array_dtype(self.iloc[:, 0]):
+            dtype = self.dtypes.iloc[0]
+            arr_type = dtype.construct_array_type()
+            values = self.values
```
IIRC the single-column case could be done without this casting.  think its worth special-casing?

TomAugspurger replied:
I don't think it can be done without casting in general. We'll still need to reshape the `(N, 1)` to `(1, N)`.

jbrockmendel replied:
im pretty sure i did this in the other PR, something like:

```
values = self._data.blocks[0].values
new_vals = [values[[n]] for n in range(len(values))]
```

(of course, if we had 2D EAs...)

TomAugspurger replied:
So, we can avoid casting to an ndarray by making ``N * P`` length-1 `__getitem__` calls, which makes `N * P` extension arrays, which are concatenated into `N` final EAs.

My prior expectation is that converting to an ndarray and doing `__getitem__` on that will be faster, and should have roughly the same amount of memory usage.

jbrockmendel replied:
what is `P` here?  in the end this probably isn't worth bikeshedding (except to add to the pile of "reasons why EAs should support 2D")

TomAugspurger replied:
Number of columns.

jbrockmendel replied:
right, i was specifically referring to single-column

jbrockmendel replied:
I'm OK with not-doing this optimization here, just want to make sure we're on the same page about what the available optimization is

TomAugspurger committed:
fixup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-transpose

TomAugspurger committed:
empty

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-transpose

jreback commented on pandas/core/frame.py:
```diff
@@ -2504,9 +2504,14 @@ def transpose(self, *args, **kwargs):
-        *args, **kwargs
-            Additional arguments and keywords have no effect but might be
-            accepted for compatibility with numpy.
+        *args : tuple, optional
```
don't we *still* need kwargs for this?

TomAugspurger replied:
AFAICT, no. Neither `np.transpose` nor `ndarray.transpose` take additional keyword arguments.

jreback commented on pandas/core/frame.py:
```diff
@@ -2504,9 +2504,14 @@ def transpose(self, *args, **kwargs):
+        *args : tuple, optional
+            Accepted for compatibility with NumPy.
+        copy : bool, default False
+            Whether to copy the data after transposing, even for DataFrames
```
versionadded tag

TomAugspurger replied:
FYI, copy isn't new. It's been around for a long time I think (the docstring in NDFrame.transpose didn't have a versionadded).

jreback commented on pandas/core/generic.py:
```diff
@@ -644,50 +644,6 @@ def _set_axis(self, axis, labels):
         self._data.set_axis(axis, labels)
         self._clear_item_cache()
 
-    def transpose(self, *args, **kwargs):
```
does Series have transpose for compat?

TomAugspurger replied:
Yes, via IndexOpsMixin.

jreback commented on pandas/core/frame.py:
```diff
@@ -2587,7 +2592,28 @@ def transpose(self, *args, **kwargs):
-        return super().transpose(1, 0, **kwargs)
+        # construct the args
+
+        dtypes = list(self.dtypes)
```
can you add a comment on what the conditions mean here

jreback commented on pandas/core/frame.py:
```diff
@@ -2587,7 +2592,28 @@ def transpose(self, *args, **kwargs):
         """
         nv.validate_transpose(args, dict())
-        return super().transpose(1, 0, **kwargs)
+        # construct the args
```
don't you think this is better located in pandas/core/reshape ? (and called as a helper function here)

TomAugspurger replied:
Maybe a slight preference for keeping it here just for readability. The reshape part is essentially just a list comprehension.

```python
values = self.values
new_values = [arr_type._from_sequence(row, dtype=dtype) for row in values]
```

which I don't think warrants its own function. I don't see anything places in `/core/reshape.py` that could use this. I believe those are reshaping to / from 1-D things. This is 2D -> 2D.

But happy to move it if you want / if you see other places that could use parts of this.

jreback reviewed (changes requested):
looks good

jbrockmendel commented on pandas/core/reshape/reshape.py:
```diff
@@ -491,6 +491,10 @@ def _unstack_extension_series(series, level, fill_value):
 
 
+def _transpose_extension_arrays(df):
+    pass
```
did this get committed on accident?

jbrockmendel commented on pandas/tests/arithmetic/conftest.py:
```diff
@@ -235,25 +235,6 @@ def box_df_fail(request):
-    # GH#23620
-    return request.param
-
-
```
nice to see this go!

jbrockmendel commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -755,10 +755,10 @@ def test_pi_sub_isub_offset(self):
 
-    def test_pi_add_offset_n_gt1(self, box_transpose_fail):
+    @pytest.mark.parametrize("transpose", [True, False])
+    def test_pi_add_offset_n_gt1(self, box, transpose):
```
i think `box` should be `box_with_array` for these

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-transpose

TomAugspurger committed:
box

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-transpose

TomAugspurger commented:
I think this is ready.

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -2586,7 +2591,29 @@ def transpose(self, *args, **kwargs):
+        else:
+            new_values = self.values.T
+            if copy:
+                new_values = new_values.copy()
```
if non-homogeneous, then `new_values` above is already a copy, can avoid re-copying here

TomAugspurger replied:
Are we 100% sure about that? Or are there types distinct dtypes that `.values` can combine without copy?

jbrockmendel replied:
if non-homogeneous then we have multiple blocks, so multiple ndarrays that are going through `np.c_` or something like it, right?  AFAIK that has to allocate new data for the output.  are there corner cases were missing @shoyer?

jbrockmendel commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -755,18 +755,18 @@ def test_pi_sub_isub_offset(self):
 
-    def test_pi_add_offset_n_gt1(self, box_transpose_fail):
+    @pytest.mark.parametrize("transpose", [True, False])
+    def test_pi_add_offset_n_gt1(self, box_with_array, transpose):
```
not a blocker, but transpose param is sub-optimal.  For DataFrame case it will correctly test both cases, but for EA/Index/Series it will mean duplicate tests.  I'll try to come up with something nicer in an upcoming arithmetic-test-specific pass

jbrockmendel commented on pandas/tests/arithmetic/test_period.py:
```diff
@@ -1011,16 +1010,15 @@ def test_parr_add_sub_td64_nat(self, box_transpose_fail):
-    def test_parr_add_sub_tdt64_nat_array(self, box_df_fail, other):
+    def test_parr_add_sub_tdt64_nat_array(self, box_with_array, other):
         # FIXME: DataFrame fails because when when operating column-wise
         #  timedelta64 entries become NaT and are treated like datetimes
```
This FIXME comment can be removed

jbrockmendel commented on pandas/tests/extension/json/test_json.py:
```diff
@@ -162,6 +162,10 @@ def test_unstack(self, data, index):
         # this matches otherwise
         return super().test_unstack(data, index)
 
+    @pytest.mark.skip(reason="Inconsistent sizes.")
```
xfail?  i.e. is this something that in principle we would like to fix some day?

jbrockmendel commented on pandas/tests/frame/test_operators.py:
```diff
@@ -891,6 +891,22 @@ def test_no_warning(self, all_arithmetic_operators):
+        df = pd.DataFrame({"a": ser, "b": ser})
+        result = df.T
+        assert (result.dtypes == ser.dtype).all()
+
```
```
round_trip = df.T.T
tm.assert_frame_equal(df, round_trip)
```
?

TomAugspurger replied:
Interestingly, the `pd.date_range("2016-04-05 04:30", periods=3).astype("category")` case fails that test. All the values are `NaT`.

I've xfalied it for now, and likely won't have time to look into it.

jbrockmendel replied:
sounds good, thanks

jbrockmendel commented:
A handful of comments, generally looks good

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into ea-transpose

TomAugspurger committed:
update test.

TomAugspurger committed:
filter

jreback commented:
@TomAugspurger merge master when you have a chance (or @jbrockmendel if this is a blocker)

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into ea-transpose

jbrockmendel committed:
fixup unused import

jbrockmendel commented:
rebased.  not a blocker for the blockwise PRs

jreback commented:
thanks @TomAugspurger 

@jbrockmendel I don't think we had a dedicated issue for this to be closed......

jbrockmendel commented:
#22120, since its about cyberpandas, i think we should ask tom to double-check that this fixes it.
